neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -222,87 +273,63 @@ def body(*args) -> BeamSearchLoopState:
-                last_word_ids=next_word_ids,
-                last_state=next_beam_prev_rnn_state,
-                last_attns=next_beam_prev_attns)
-
-            # For run-time computation, the decoder needs:
-            # - step
-            # - input_symbol
-            # - prev_rnn_state
-            # - prev_rnn_output
-            # - prev_contexts
-            # - attention_loop_states
-            # - finished
-
-            # For train-mode computation, it also needs
-            # - train_inputs
-
-            # For recording the computation in time, it needs
-            # - rnn_outputs (TA)
-            # - logits (TA)
-            # - mask (TA)
-
-            # Because of the beam search algorithm, it outputs
-            # (but does not not need)
-            # - prev_logits
+                                   len(self.vocabulary))
+
+            next_feedables_dict = {"input_symbol": next_word_ids}
+            for key, val in dec_loop_state.feedables._asdict().items():
+                if key == "step" or key == "input_symbol":
+                    continue
+
+                if isinstance(val, tf.Tensor):
+                    next_feedables_dict[key] = tf.gather(val, next_beam_ids)
+                elif isinstance(val, list):
+                    if not all(isinstance(t, tf.Tensor) for t in val):
+                        raise TypeError("Expected tf.Tensor among feedables")
+
+                    next_feedables_dict[key] = [tf.gather(t, next_beam_ids)
+                                                for t in val]
+                else:
+                    raise TypeError("Expected only tensors or list of tensors "
+                                    "among feedables")
+
+            next_beam_lengths = tf.gather(hyp_lengths, next_beam_ids)
 
             # During beam search decoding, we are not interested in recording
             # of the computation as done by the decoder. The record is stored
             # in search states and step outputs of this decoder.
+            next_feedables = dec_loop_state.feedables._replace(
+                **next_feedables_dict)
+
+            dec_loop_state = dec_loop_state._replace(feedables=next_feedables)
 
-            next_prev_logits = tf.gather(next_loop_state.prev_logits,
-                                         next_beam_ids)
+            # CALL THE DECODER BODY FUNCTION
+            # TODO figure out why mypy throws too-many-arguments on this
+            next_loop_state = decoder_body(*dec_loop_state)  # type: ignore
 
-            next_prev_contexts = [tf.gather(ctx, next_beam_ids) for ctx in
-                                  next_loop_state.prev_contexts]
+            next_search_state = SearchState(
+                logprob_sum=next_beam_logprob_sum,
+                prev_logprobs=tf.nn.log_softmax(
+                    next_loop_state.feedables.prev_logits),
+                lengths=next_beam_lengths,
+                finished=dec_loop_state.feedables.finished)
```

Comment:
SearchState.finished nemuzes vytvaret na zaklade tf.gather(dec_loop_state.feedables.finished, next_beam_ids).

Vezmi napriklad situaci, ze v jednom kroku 2x picknes beam, ktery mel "<EOS>" jako symbol s nejvetsim logprobem (tim padem, je oznaceny jako finished). Jenze druhy nejpravdepodobnejsi symbol uz neni "<EOS>", takze ten by naopak nemel byt oznaceny jako finished.

Navic dec_loop_state.feedables.finished obsahuje hodnoty ktere vznikaji pred ensemblovanim.

Reply:
Taky jsem si to myslel, ale pak jsem si řekl následující: ten seznam příznaků finished je ke každé hypotéze jeden, takže když tady volám `tf.gather` na finished podle toho, jaké vybírám hypotézy k pokračování, tak s každou hypotézou už pokračuje dál, jestli je finished a jedno, jestli, kolikátá byla v pořadí. Nebo ne?

Reply:
No jestli se nepletu, tak to tedka běží takto:
1. Spustí se decoder_body. To vygeneruje logity, vybere next_symbol (podle flagů: train, sample nebo argmax) a na základe next_symbol updatuje finished.

[Začátek iterace tf.while_loop]
(Zde je místo, kde se počítají logproby, které v runneru ensemblujeme a zpětně feedujeme.)

2. Na základě logprobů se spočítají beam_scores a vybere se top_k nových hypotéz. Některé tyto hypotézy ale mohou mít odlišný next_symbol než byl vybraný v 1. a je tedy přepočítat finished flagy pro tyto hypotézy.

3. S novýma hypotézama opět spouštíme decoder_body.
[Konec iterace tf.while_loop]

Myslím, že by se tam tedy měl vrátit to počítání "finished" pro top_k hypotézy.

Reply:
Jo, ale ten gather tam nikde není koukám.

Reply:
https://github.com/ufal/neuralmonkey/blob/4db2be8fd6066b5b85d5bb2543a02c24197832cb/neuralmonkey/decoders/beam_search_decoder.py#L309

Poté, co se vytvoří next_feedables, tak by se mělo updatovat next_feedables.finished podle hodnot next_feedables.input_symbol.